### PR TITLE
Wallet - Add malli spec to Quo Wallet components (batch 1) #18354

### LINKED ISSUE #18354: Wallet - Add malli spec to Quo Wallet components (batch 1)
To stabilise the wallet project we want to add malli specs to some of the more important/frequently used UI components in the wallet flows.

Add a Malli spec for the following components ->

**Quo / Wallet**
- account-card 
- account-origin 
-  account-overview 
- account-permissions 
- address-text
-  confirmation-progress 

For more info see this pr: https://github.com/status-im/status-mobile/pull/17867

Some examples can be found already in the codebase by searching for uses of namespace schema.core

ping @yqrashawn, @clauxx or @ilmotta if any help is needed when writing these specs.

### PR DESCRIPTION
**Pull request by mmilad75**
fixes #18354 

The following components are updated:

**Quo / Wallet**
- account-card
- account-origin
- account-overview
- account-permissions
- address-text
- confirmation-progress

### CODE REVIEW

File: src/quo/components/wallet/account_card/schema.cljs [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+
+(def ?base
+  [:map
+   [:type {:optional true} [:enum :default :watch-only :add-account :empty :missing-keypair]]
+   [:customization-color {:optional true} [:maybe :schema.common/customization-color]]
+   [:theme :schema.common/theme]
+   [:metrics? {:optional true} [:maybe :boolean]]
+   [:on-press {:optional true} [:maybe fn?]]])
+
+(def ?amount
```
Comment (J-Son89) [inline src/quo/components/wallet/account_card/schema.cljs]: I think we should mark internal schemas private

Comment (J-Son89) [reply src/quo/components/wallet/account_card/schema.cljs]: I like how you split up the schemas 🙌

File: src/quo/components/wallet/account_overview/view.cljs [hunk 2]
```diff
@@ -2,10 +2,12 @@
   (:require
     [quo.components.icon :as icons]
     [quo.components.markdown.text :as text]
+    [quo.components.wallet.account-overview.schema :as component-schema]
     [quo.components.wallet.account-overview.style :as style]
     [quo.foundations.colors :as colors]
     [quo.theme :as quo.theme]
     [react-native.core :as rn]
+    [schema.core :as schema]
```
Comment (J-Son89) [inline src/quo/components/wallet/account_overview/view.cljs:10]: imo we should move this schema.core stuff to schema files and create the wrapper in there. We should only need one import into a view file related to schemas.
wdyt @ilmotta, @clauxx ?

File: src/status_im/contexts/preview/quo/wallet/transaction_progress.cljs [hunk 3]
```diff
@@ -142,8 +142,8 @@
                 :epoch-number-mainnet         "181,329"
                 :epoch-number-optimism        "181,329"
                 :epoch-number-arbitrum        "181,329"
-                :optimism-progress-percentage "10"
```
Comment (J-Son89) [inline src/status_im/contexts/preview/quo/wallet/transaction_progress.cljs:145]: curious, is this a result of adding the schema? :)

Comment (mmilad75) [reply src/status_im/contexts/preview/quo/wallet/transaction_progress.cljs:145]: Yes, some use cases were wrong and I fixed them

File: src/quo/components/wallet/account_permissions/schema.cljs [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+  [:=>
+   [:catn
+    [:props
+     [:map
+      [:account
+       [:map
+        [:name [:maybe :string]]
+        [:address [:maybe :string]]
+        [:emoji [:maybe :string]]
+        [:customization-color [:maybe [:or :string :keyword]]]]]
```
Comment (clauxx) [inline src/quo/components/wallet/account_permissions/schema.cljs]: Are these keys always going to be passed and sometimes may be (explicitly) `nil` or are they optional? If so, should add `{:optional true}`

File: src/status_im/contexts/wallet/home/view.cljs [hunk 5]
```diff
@@ -60,7 +60,7 @@
            :data                              cards
            :horizontal                        true
            :separator                         [rn/view {:style style/separator}]
-           :render-fn                         quo/account-card
+           :render-fn                         (fn [item] [quo/account-card item])
```
Comment (clauxx) [inline src/status_im/contexts/wallet/home/view.cljs:63]: why is this needed?

File: src/quo/components/wallet/account_overview/schema.cljs [hunk 6]
```diff
@@ -0,0 +1,20 @@
+(ns quo.components.wallet.account-overview.schema)
+
+(def ?schema
+  [:=>
+   [:catn
+    [:props
+     [:map
```
Comment (clauxx) [inline src/quo/components/wallet/account_overview/schema.cljs:7]: Since we didn't yet arrive at a decision on open/close maps this is not critical, but since this component's props are simpler I'd suggest we close the map to keep it consistent with the rest of the schemas.

**Review by clauxx — APPROVED** [review]
Great work with malli 💯! Looks good aside from some small comments

File: src/status_im/contexts/wallet/home/view.cljs (hunk 5 above)
Comment (mmilad75) [reply src/status_im/contexts/wallet/home/view.cljs:63]: It was throwing malli errors before this. I don't know why 🤔

File: src/quo/components/wallet/account_overview/schema.cljs (hunk 6 above)
Comment (mmilad75) [reply src/quo/components/wallet/account_overview/schema.cljs:7]: This one is used in some places that receives data directly from backend, so a lot of extra props are passed to it. It's better to let it be open for now

**Review by J-Son89 — APPROVED** [review]

File: src/quo/components/wallet/account_overview/schema.cljs [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+      [:account-name {:optional true} [:maybe :string]]
+      [:currency-change {:optional true} [:maybe :string]]
+      [:current-value {:optional true} [:maybe :string]]
+      [:percentage-change {:optional true} [:maybe :string]]
+      [:time-frame-string {:optional true} [:maybe :string]]
```
Comment (ilmotta) [inline src/quo/components/wallet/account_overview/schema.cljs:12]: Folks, I've been thinking, it would be good if we arrived at a guideline for predicates in schemas sooner rather than later.

Malli accepts all predicates to be functions or keywords (if they've been added to the registry). The benefit of predicates as functions is that we get compile time errors from the CLJS compiler. Many times, I made typos like `:sting` instead of `:string`, but this will cause a runtime error later on or no error at all if there's no test code exercising the schema (which would lead to merging invalid schemas sometimes).

It's nice to be able to represent schemas as data. They could be easily serialized, but we don't use this feature.

Therefore, I'm inclined to prefer functions as predicates, not keywords. Instead of `:string`, it would be `string?`. It's just as terse, but with compile-time protection.

Do you have another perspective? Or does this sound reasonable for a guideline in `schema/README`?

cc @J-Son89, @clauxx, @shivekkhurana

File: src/status_im/contexts/wallet/home/view.cljs (hunk 5 above)
Comment (ilmotta) [reply src/status_im/contexts/wallet/home/view.cljs:63]: We should probably understand why because we should be able to just pass `quo/account-card` without wrapping.

File: src/quo/components/wallet/account_overview/view.cljs (hunk 2 above)
Comment (clauxx) [reply src/quo/components/wallet/account_overview/view.cljs:10]: It could be a function that would wrap instrumentation, but it feels boilerplate-ish to have in each. Would be easier if we used `schema/=>`, then from the component-schema file we could just require the `view` and instrument it, but there was an issue with that that I don't remember now from our initial conversations.

Comment (ilmotta) [reply src/quo/components/wallet/account_overview/view.cljs:10]: I don't quite understand the problem of requiring `schema.core`, since that's the namespace exposing `instrument`.

File: src/quo/components/wallet/account_permissions/schema.cljs (hunk 4 above)
Comment (mmilad75) [reply src/quo/components/wallet/account_permissions/schema.cljs]: Done.

File: src/quo/components/wallet/account_overview/schema.cljs (hunk 6 above)
Comment (shivekkhurana) [reply src/quo/components/wallet/account_overview/schema.cljs:7]: +1 for open. 

1. Backend passes props that the component might not need
2. Quo preview passes `@state` while calling the component, which may contain extra kv pairs.

File: src/quo/components/wallet/account_card/schema.cljs (hunk 1 above)
Comment (mmilad75) [reply src/quo/components/wallet/account_card/schema.cljs]: Done

File: src/quo/components/wallet/account_permissions/schema.cljs [hunk 8]
```diff
@@ -0,0 +1,22 @@
+(ns quo.components.wallet.account-permissions.schema
+  (:require [quo.components.wallet.required-tokens.view :as required-tokens]))
+
+(def ?schema
+  [:=>
+   [:catn
+    [:props
+     [:map
+      [:account
```
Comment (shivekkhurana) [inline src/quo/components/wallet/account_permissions/schema.cljs:9]: Do you see value in extracting `account` to `schema.common` ?

File: src/quo/components/wallet/account_overview/schema.cljs [hunk 9]
```diff
@@ -0,0 +1,20 @@
+(ns quo.components.wallet.account-overview.schema)
+
+(def ?schema
+  [:=>
+   [:catn
+    [:props
+     [:map
+      [:account-name {:optional true} [:maybe :string]]
+      [:currency-change {:optional true} [:maybe :string]]
+      [:current-value {:optional true} [:maybe :string]]
```
Comment (shivekkhurana) [inline src/quo/components/wallet/account_overview/schema.cljs:10]: Shouldn't this be `:number` ?

**Review by shivekkhurana — APPROVED** [review]
Looks good overall, marked some lines with my doubts.

File: src/quo/components/wallet/account_origin/schema.cljs [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+   [:customization-color {:optional true} [:maybe :schema.common/customization-color]]])
+
+(def ^:private ?recovery-phrase
+  [:map
+   [:derivation-path {:optional true} [:maybe :string]]
+   [:on-press {:optional true} [:maybe fn?]]])
+
+(def ?schema
+  [:=>
+   [:catn
```
Comment (ilmotta) [inline src/quo/components/wallet/account_origin/schema.cljs:24]: Not a suggestion for change @mmilad75, but rather a perception to incite possible guidelines.

Maybe we should prefer `cat` (not `catn`) for function schemas with just one argument, as is the case for most components. It adds an extra indentation and when these schemas fail it will be just as clear why. One might say it's nice for consistency because `catn` leads to more readable errors, but to me it seems superfluous to use it.

Maybe the best is to just keep the option so the developer use whatever they prefer, but we shouldn't feel forced to use `catn` all the time IMO.

File: src/quo/components/wallet/account_permissions/schema.cljs (hunk 8 above)
Comment (ilmotta) [reply src/quo/components/wallet/account_permissions/schema.cljs:9]: Where to put schemas is always difficult to decide. From one angle, I prefer to think `schema.common` is the place for schemas more decoupled from domains. Something where we would describe a big integer, a public key, a theme, a style map, and so on. But an account schema could live outside `schema.common`. This would help us separate things based on bounded contexts. Otherwise there's a tendency `common` stuff can absorb anything.

An account can mean other things besides a wallet account. Maybe if we want to define a schema in the global registry, it should be in a different namespace, so the keyword in the registry would be `:schema.wallet/account`.

Or maybe we shouldn't couple quo components with domain concepts, even though they look like they depend on them. This would lead to a bit of duplication of schemas, but then we would be able to evolve schemas independently from quo. Just a tradeoff.

Long train of thought, just ideas on the table @shivekkhurana.

File: src/quo/components/wallet/account_overview/schema.cljs (hunk 9 above)
Comment (ilmotta) [reply src/quo/components/wallet/account_overview/schema.cljs:10]: Good point. `:string` is too loose to represent numbers. For instance, if the value is a string because it's an encoded big integer, then it would be better to at least create an alias to represent this, e.g. `:schema.common/raw-bigint`. I'm not sure if that's the case here, but the `:string` schema is also a good example of a type that doesn't help me figure out how to deal with it.

In the future, we could also consider only passing actual numbers throughout the app, even if some of them are representations of big numbers and need to be decoded as such. I did this in the past and it can bring sanity to the whole system. But we would need to make sure all numbers are decoded (from status-go responses) when written to the app-db.

File: src/status_im/contexts/wallet/home/view.cljs (hunk 5 above)
Comment (ulisesmac) [reply src/status_im/contexts/wallet/home/view.cljs:63]: @ilmotta The original flat-list implementation expects a function to be passed.

Just saying again that we should revisit our wrapper :eye:

Comment (ilmotta) [reply src/status_im/contexts/wallet/home/view.cljs:63]: But `quo/account-card` is a function, so it surprises me that we need to wrap it in a dummy closure.

Comment (ilmotta) [reply src/status_im/contexts/wallet/home/view.cljs:63]: > Just saying again that we should revisit our wrapper 👁️

Just saying... I see you! Yes, that wrapper function seems very inefficient with the creation of anonymous functions over and over.

File: src/quo/components/wallet/account_permissions/schema.cljs (hunk 8 above)
Comment (shivekkhurana) [reply src/quo/components/wallet/account_permissions/schema.cljs:9]: There are two schools of typings that I have observed:
1. Move all types to their own ns
This helps avoid circular deps. that may arise in future

2. Keep types close to implementation
Better docs, easier to implement as no extra files are needed, available at the point of execution.

I personally prefer 2 (ie local types), but there are some domain models that will be repeated. `Account` being one of them maybe?

So in my head, I assumed `schema.common` to absorb any type that appears more than once.

File: src/quo/components/wallet/account_overview/schema.cljs (hunk 7 above)
Comment (J-Son89) [reply src/quo/components/wallet/account_overview/schema.cljs:12]: sounds good to me 👍

File: src/quo/components/wallet/account_overview/view.cljs (hunk 2 above)
Comment (J-Son89) [reply src/quo/components/wallet/account_overview/view.cljs:10]: https://www.youtube.com/watch?v=_K-L9uhsBLM :)

File: src/quo/components/wallet/account_overview/view.cljs [hunk 11]
```diff
@@ -129,16 +131,5 @@
...
-  | `:time-frame`        | A value that represents the type of the timeframe, Can be from a preset of time frames. `[:1-week :1-month :3-months :1-year :custom]` (default `nil`) if custom is set, We expect a start time and if there's a space between two times that are in `time-frame-string` we'll split them with an arrow to the right icon
-  | `:loading?`          | A value that indicates that component is loading data. (default `nil`)
-  | `:metrics`           | A value that indicates if the account value have increased can be `:positive` or `:negative` (default `nil`)
-  | `:time-frame-string` | A value representing time frame string. (default `nil`)
-  | `:current-value`     | A value representing the current value of the selected account. (default `nil`)
-  | `:currency-change`   | A value representing the change of the value of the selected account in the selected currency. (default `nil`)
-  | `:percentage-change` | A value representing the change of the value of the selected account in percentage relative to it's previous value. (default `nil`)"
-  (quo.theme/with-theme view-internal))
+  (quo.theme/with-theme
```
Comment (J-Son89) [inline src/quo/components/wallet/account_overview/view.cljs:134]: 🙌

File: src/quo/components/wallet/account_permissions/schema.cljs (hunk 8 above)
Comment (ilmotta) [reply src/quo/components/wallet/account_permissions/schema.cljs:9]: > to absorb any type that appears more than once.

I also prefer option 2 @shivekkhurana whenever possible and only use the global registry if really necessary.

Maybe the word `common` is not the best word when I picked it up, because it implies any schema common to two or more namespaces can already live up there. We're already starting to have discussions about having schemas for data crossing the status-go <-> mobile barrier (e.g. in contract tests and events), which means it won't take long until we have lots of _common_ schemas (complicated ones, not like quo schemas) that are all representing entities that are not tied to any particular part of the app.

Putting them all under the keyword `:schema.common/*` can work for a while, but something tells me we'll refactor later on and break them apart by domains, so instead of `:schema.common/wallet-account` we would have `:schema.wallet/account`. Of course, I'm guessing about this supposed future, we may never get there and so using the `:schema.common/` qualification is totally fine now.

File: src/quo/components/wallet/account_overview/view.cljs (hunk 2 above)
Comment (ilmotta) [reply src/quo/components/wallet/account_overview/view.cljs:10]: hauahahaha any small time make big time!

**Review by ilmotta — APPROVED** [review]

**Comment by mmilad75** [comment]
@status-im/mobile-qa Please check the e2e result

**Comment by qoqobolo** [comment]
> @status-im/mobile-qa Please check the e2e result

Hey @mmilad75, thanks for the PR! We are experiencing the data delivery issue, which is the reason for most of the e2e failures. They are not related to the PR.
